binding.c

```diff
@@ -309,254 +392,294 @@ on_udx_stream_read (udx_stream_t *stream, ssize_t read_len, const uv_buf_t *buf)
     }
   }
 
-  napi_env env = n->env;
+  js_env_t *env = n->env;
 
-  napi_handle_scope scope;
-  napi_open_handle_scope(env, &scope);
+  js_handle_scope_t *scope;
+  int err = js_open_handle_scope(env, &scope);
+  assert(err == 0);
 
-  napi_value ctx;
-  napi_get_reference_value(env, n->ctx, &ctx);
+  js_value_t *ctx;
+  err = js_get_reference_value(env, n->ctx, &ctx);
+  assert(err == 0);
 
-  napi_value callback;
-  napi_get_reference_value(env, n->on_data, &callback);
+  js_value_t *callback;
+  err = js_get_reference_value(env, n->on_data, &callback);
+  assert(err == 0);
 
-  napi_value argv[1];
-  napi_create_uint32(env, read, &(argv[0]));
+  js_value_t *argv[1];
+  err = js_create_uint32(env, read, &(argv[0]));
+  assert(err == 0);
 
-  napi_value res;
+  js_value_t *res;
 
-  if (napi_make_callback(env, NULL, ctx, callback, 1, argv, &res) == napi_pending_exception) {
-    napi_value fatal_exception;
-    napi_get_and_clear_last_exception(env, &fatal_exception);
-    napi_fatal_exception(env, fatal_exception);
+  err = js_call_function(env, ctx, callback, 1, argv, &res);
```

Comment:
This should do a microtask checkpoint assuming that `on_udx_stream_read()` is an I/O callback.